src/field/gf_101_2.rs

```diff
@@ -0,0 +1,335 @@
+use core::{
+  iter::{Product, Sum},
+  ops::{Add, AddAssign, Mul, Neg, Sub, SubAssign},
+};
+use std::ops::{Div, DivAssign, MulAssign, Rem};
+
+use serde::{Deserialize, Serialize};
+
+use super::ExtensionField;
+use crate::field::FiniteField;
+
+/// Pluto curve with modulus 101 supports two degree extension field. This can be verified
+/// by finding out embedding degree of the curve, i.e. smallest k: r|q^k-1
+const EXT_DEGREE: usize = 2;
+const QUADRATIC_EXTENSION_FIELD_ORDER: u32 = 101 * 101;
+
+#[derive(Clone, Default, Copy, Eq, PartialEq, Hash, Debug, Serialize, Deserialize)]
+
+/// Quadratic Extension field element represented as polynomial of degree 1 in form:
+/// a_0 + a_1*t where {a_0, a_1} \in \mathhbb{F}. Uses irreducible poly of the form:
+/// (X^2-K).
+pub struct QuadraticPlutoField<F: FiniteField> {
+  pub(crate) value: [F; 2],
+}
+
+impl<F: FiniteField> QuadraticPlutoField<F> {
+  const D: usize = 2;
+
+  /// irreducible polynomial used to reduce field polynomials to second degree:
+  /// F[X]/(X^2-2)
+  fn irreducible() -> F { F::from_canonical_u32(2) }
+}
+
+impl<F: FiniteField> ExtensionField<F> for QuadraticPlutoField<F> {
+  const D: usize = EXT_DEGREE;
+
+  fn irreducible() -> F { Self::irreducible() }
+
+  fn from_base(b: F) -> Self { Self { value: [b, F::zero()] } }
+}
+
+impl<F: FiniteField> From<F> for QuadraticPlutoField<F> {
+  fn from(value: F) -> Self { Self::from_base(value) }
+}
+
+impl<F: FiniteField> FiniteField for QuadraticPlutoField<F> {
+  type Storage = u32;
+
+  const ORDER: Self::Storage = QUADRATIC_EXTENSION_FIELD_ORDER;
+
+  fn zero() -> Self { Self { value: [F::zero(); EXT_DEGREE] } }
+
+  fn one() -> Self { Self { value: [F::one(), F::zero()] } }
+
+  fn two() -> Self { Self { value: [F::two(), F::zero()] } }
+
+  fn neg_one() -> Self { Self { value: [F::neg_one(), F::zero()] } }
+
+  // field generator: can be verified using sage script
+  // ```sage
+  // F = GF(101)
+  // Ft.<t> = F[]
+  // P = Ft(t ^ 2 - 2)
+  // F_2 = GF(101 ^ 2, name="t", modulus=P)
+  // f_2_primitive_element = F_2.primitive_element()
+  // ```
+  fn generator() -> Self {
+    // TODO: unsure if this is correct or not, research more
+    Self { value: [F::from_canonical_u32(15), F::from_canonical_u32(20)] }
+  }
+
+  /// Computes the multiplicative inverse of `a`, i.e. 1 / (a0 + a1 * t).
+  /// Multiply by `a0 - a1 * t` in numerator and denominator.
+  /// Denominator equals `(a0 + a1 * t) * (a0 - a1 * t) = a0.pow(2) - a1.pow(2) * Q::residue()`
+  fn inverse(&self) -> Option<Self> {
+    if *self == Self::zero() {
+      return None;
+    }
+
+    let mut res = Self::default();
+    let scalar =
+      (self.value[0].square() - Self::irreducible() * self.value[1].square()).inverse().unwrap();
+    res.value[0] = self.value[0] * scalar;
+    res.value[1] = -self.value[1] * scalar;
+    Some(res)
+  }
+
+  #[inline]
+  fn pow(&self, power: Self::Storage) -> Self {
+    // TODO: not ideal to duplicate across traits impls
```

Comment:
Maybe we can add a default impl in the trait what do you think?

Reply:
Looking good!

Reply:
done